WIP Refactor - Session Handling

dr7ana opened the pull request:
## Inheritence Pre-Refactor:
- Pathset
  - Builder
    - service::Endpoint
      - TunEndpoint
      - NullEndpoint
    - BaseSession
      - SNodeSession
      - ExitSession
    - OutboundContext

## Inheritence Post-Refactor:
- PathBuilder
  - service::Handler = Remote service management
    - TunEndpoint (extending handlers::BaseHandler)
    - NullEndpoint (extending handlers::BaseHandler)
  - service::Endpoint = Local service management
  - session::BaseSession
    - session::ClientSession = Service provided by a client
    - session::SNodeSession = Service provided by a SNode via lokinet
    - session::ExitSession = Exit provided by a client
  - exit::handler = Remote exit node management
  - exit::Endpoint = local exit node management

### Base Classes:
  - EndpointBase
    - maintain session information
      - storing convotag chosen by session initiator
  - PathBuilder
    - building paths
      - Local: to random pivots
      - Remote: to specific pivots
    - tracking available paths/currently selected path for session

### Derived Classes:
  - service::Endpoint = Local Service Management
  - service::Handler = Remote Service Management
  - exit::Endpoint = Local Exit Node Management
  - exit::Handler = Remote Exit Node Management

### Auxiliary Classes:
  - handlers::BaseHandler = code specific to {Tun,Null}Endpoint
  - auth::AuthPolicy = auth handling for local/remote exit management

dr7ana committed:
Squashed commits for merge errors afer fixing client-refactor:

- Deprecate pathset, smashed into PathBuilder (renamed to PathHandler)
- Re-abstraction of sessions and PathHandlers
  - Renamed PathBuilder -> PathHandler to more accurately reflect purpose
  - {Service,Exit}Handler will remain as PathHandlers, though currently no path-sharing amongst sessions being managed is to be implemented. Handlers will maintain their own paths for both lookups and initiating sessions, while sessions will manage their paths independantly.
  - Session object handling necessitates the differentiation between outbound and inbound sessions. Initiators of sessions are entirely responsible for dictating the chosen path on which the session communicates, and must therefore continually build and manage paths for the negotiated session.
  - Outbound sessions are now {Service,Exit}Sessions
  - Inbound sessions are implemented with InboundSession, which is agnostic to the type of service being operated locally (service vs exit, client vs relay, etc). When the Session initiator signals a switch to a different path, it will be assigned to the InboundSession object by {Service,Exit}Endpoint, which manages local services and exits

jagerman commented:
- `session::ExitSession` should be a subclass of `session::ClientSession` (because it's still a client session, but with "extra" stuff for exit traffic handling)

dr7ana committed:
libquic testing out verbose bparser stuff

dr7ana committed:
deprecated byte_t (twas a dumb typedef anyways)

dr7ana committed:
ToString is no longer camelcase because why would it be???

dr7ana committed:
deprecated old types from config

dr7ana committed:
bump libquic for updated exposed ranges

dr7ana committed:
deprecated net/ip_address and path/path_context

dr7ana committed:
bump libquic to expose ev loop

dr7ana committed:
event loop swap

- brought up libev loop implementation from libquic
- this will streamline deprecating the old IP_packet, UDPHandle, and other types

dr7ana committed:
link_manager generates static_secret for quic ep

dr7ana committed:
deprecation nation

- old types almost entirely deprecated (ip_packet, sockaddr, etc)
- need to touch base with the team to discuss packet routing and the like

dr7ana committed:
libquic bump

dr7ana committed:
compiling, not linking

dr7ana committed:
C++ version, clang-format

- C++ bumped 17 -> 20
- clang-format (plus all clangs) -> 17

dr7ana committed:
libquic bump

dr7ana committed:
bye bye bencode baby bye byeeee

dr7ana committed:
fully deprecated old logging

dr7ana committed:
libquic bump

dr7ana committed:
ip packet checksums for ipv4/6, udp/tcp

dr7ana committed:
templated remoteaddress object

- eliminated {client,relay}address
- better loops in new key types

dr7ana committed:
config parsing cleaned up for owned ranges and addresses

dr7ana committed:
compile fixes (except for consteval fmt::format oddities

dr7ana committed:
re-separated address classes

- there is one key difference between the address classes: ClientAddress can be constructed with an ONS name, and can therefore exist in one of 2 dual states (pubkey vs ONS)

dr7ana committed:
Putting new objects to use

- cleaned up config parsing using the better Address/IP types and quic::Address where appropriate
- created new templated address_map using concepts constraining local types to quic::Address and IPRange, and remote types to {Client,Relay}Address
- now the correct local addresses and ranges can be passed to the correct handlers and endpoints for remote and local exit/service operation
- basically fully configured ::RemoteHandler
- deleted deprecated object files previously removed from compilation

dr7ana committed:
Restructuring, session initiation

- combined {exit,service}::Endpoint into handlers::LocalEndpoint
- combined {exit,service}::Handler into handlers::RemoteHandler
- BaseSession subsumed {Client,Relay}Session -> OutboundSession
- Session initialization code underway

dr7ana committed:
libquic c++20/ev-loop/ip bump

dr7ana committed:
Transposed address types

- It became increasingly tedious to partition the address types between {Client,Relay}Address, as a client can operate an exit or hidden service while a relay could also be operating a hidden service. The client's hidden service would terminate in a `.loki` while the relay's would terminate in a `.snode`, complicating the ONS logic implemented in ClientAddress
- Address types were partitioned into {Network,Relay}Address; NetworkAddress conceptually encapsulates all remote hidden services (agnostic of client vs relay) and exit nodes being operated on the network: RelayAddress specifically designates an addressible relay to be used in scopes excluding hidden services and exit nodes

dr7ana committed:
Design review meeting

- Session initiation code laid out from path build to successful session object creation
- NullEndpoint renamed to more apt descriptor

dr7ana committed:
libquic vbump

dr7ana committed:
Segmented path-build

- Path building logic in ::PathHandler is broken up into 3 functions. This allows paths built outside of the usual PathHandler::Tick() logic (ex: paths for initiating sessions) to implement custom logic and path handling
- RemoteHandler now uses segmented build logic to create a path upon session initiation without storing the created path internally

dr7ana committed:
Removed outer hash for path messages

- if a hop can generate the symmetric key and decrypt its respective payload, then the concept of checking a hash is somewhat pointless
- hmac generation and validation removed

dr7ana committed:
DRY'ed out symmetric encryption logic in path-build and path-controls

dr7ana committed:
Symmetric encryption

- DRYed out symm. enc. in path-building and session initiation
- onwards to unifying session initiation/path-build into one-shot

dr7ana committed:
Session init

- correctly looks up introset locally, stores on successful request-response if not found locally
- fetched introset used for recursive path-build through pivots
- serialize/deserialize logic DRYs out implementation
- session init handling laid out

dr7ana committed:
Router prefiguring, ons resolve

- LocalEndpoint and RemoteHandler properly configured using values from ::NetworkConfig
- ONS range mappings passed to config now resolved by RemoteHandler at earliest possible phase in Router::run()

dr7ana committed:
libquic vbump

dr7ana committed:
Make it reachable

- LocalEndpoint logic for regenerating and publishing introset
- 'publish_intro' performed over path messages: to be revised upon evaluating determinism in dht key comparison with every relay holding the entire network's RC's at all time
- surrounding logic left unchanged to be revised as per previous bullet -- can also likely deprecate "order" field with more massaging
- futher deprecation of "is_relayed" bool field would require different endpoints for relays to call

dr7ana committed:
Prefiguring, deprecation

- TunEndpoint and LocalEndpoint prefiguring
- Fully removed service::Address and all uses of AddressVariant_t in favor of the improved NetworkAddress type
- Fully removed exit/* and most of service/*

dr7ana committed:
Buffer methods

- Fixed final bencode.h dependencies in AlignedBuffer that was pending refactor of other types
- Last traces of any non-oxenc bencode are done
- Next is TCP listeners!

dr7ana committed:
TCP listener laid out

- Was able to repurpose a lot of the code from the ill-fated ev-dns implementation
- TCP{Handle,Socket} implemented to parallelize UDP{Handle,Socket} in functionality
- TCP implementation takes in a callback that receives an IPPacket; can also be easily reconfigured to hanle a UDPPacket by simply changing the callback type and the constructor called in llarp/ev/tcp.cpp::tcp_read_cb()

dr7ana committed:
Netif and quic Tunnel reduction

- link::TunnelManager will be re-implemented in link::QUICTunnel to make and accept TCP connections
- Will be simplified using Libevent and hooked into the new LocalEndpoint/RemoteHandler model
- TCP{Handle,Socket} ownership model will be revised to more monadic model; this will allow the TCPSockets (sockets representing accepted connections) to function entirely independantly of the bound socket

dr7ana committed:
libquic vbump

- manual routing
- Network ctor from existing Loop

dr7ana committed:
So much

- EmbeddedEndpoint removed for solo implementation of TunEndpoint for all instances except non-embedded clients
- Deprecated TransitHopInfo, TransitHopID; absorbed into TransitHop
- Reworked PathContext internal storage of TransitHops, both mutators/setters/getters
- Straightened out success conditions of path building
- Session initiation and path respective path building code is almost done being laid out
- Full steam ahead buddy

dr7ana committed:
Session initation, authentication

- Session initiation/handling session initiation request logic finished; still to loop in TCP logic and handlers
- remove SessionAuthPolicy from OutboundSession
- auth token mapping passed to RemoteHandler; ONS auth mappings resolved with ONS range mappings
- auth_static_tokens passed to LocalEndpoint; received auth verify passed tokens against that

dr7ana committed:
TCP client code hooked up

- TCP handles and connections are now looped into session initiation and reception
- I'm tired

dr7ana committed:
TCP tunnel differentiated inbound/outbound

- TCP backend is differentiated in Inbound vs Outbound sessions, as one listens while the other connects
- These pathways are mroe discretely implemented, and the logic may move into the derived classes instead
- Now to add hooks to connect TCP server/client to backend (app or system TUN)

dr7ana committed:
Client session refactor, datagram pipelining

- LocalEndpoint/RemoteHandler combined into SessionEndpoint for ambidextrous session management
- Data message routing in/out of paths implemented using callbacks for receiving datagrams differently for embedded vs full clients
- Tunneled QUIC endpoints (no-UDP endpoints) implemented with TCP client/server endpoints on embedded/full clients
- Lots of surrounding work on the above points

dr7ana committed:
Dep bump + misc

- libquic bumped for needed ip methods
- bumped nlohmann, cpr, CLI11, oxen-encoding, pybind11, sqlite_orm
- span-lite, ghc-filesystem, and uvw removed after shift to C++20

dr7ana committed:
IPRange iterator

- Tunnel IP assignment requires robust IP assignment
- With the correct implementation of max_ip/next_ip retrieval from ip and ip_range objects, we can now implement IPRangeIterator to manage this

dr7ana committed:
TUN/Session IP routing negotiation

- TUN endpoint assigns a local IP to route traffic for inbound/outbound sessions
- Config mappings pre-loaded and superside ip iteration

dr7ana committed:
TUN ip packet re-writing; deps bump

- Packets coming in and out of full clients need their IP src/dest addresses re-written
- Lokinet's TUN device sucks up all packets aimed at the if range, correctly addressing them to the proper outbound
- On reception, lokinet's TUN device addresses correctly for the mapped application receiving the packet
- Brought a lil somethin' somethin' from libquic for CI
- Bump oxenmq to cpp20 branch
- Revert cpr to 1.9 stable for subproject curl dep bug

dr7ana committed:
CI fixes and architecture changes; bt serialize path object fix; deps bumps

dr7ana committed:
testing something on CI

dr7ana committed:
libquic bump -> dr7ana/repeater

dr7ana committed:
it's... alive!?

dr7ana committed:
libquic bump

dr7ana committed:
tick logic handling

- libevent loop ticker now handles all execution of `Router::tick()`

dr7ana committed:
Aggregated bugfix commits and logging fixes

- Formatting concept to enable oxen-logging printing for custom objects was updated with libquic vbump to require boolean member variable
- Members added in base objects to be inherited through

dr7ana committed:
libquic -> dr7ana/primary_net_closure branch

dr7ana committed:
ipv6 config disabled, tun interface solid on ipv4

dr7ana committed:
kick testnet buddy

dr7ana committed:
so much better!

dr7ana committed:
Misc (squash)

dr7ana committed:
kick

dr7ana committed:
issue fixed, force clearing nodedb

dr7ana committed:
better netif

dr7ana committed:
libevent 2.2 for watchers

dr7ana committed:
more netif

dr7ana committed:
testing this out

dr7ana committed:
libquic -> dr7ana/0rtt, etc

dr7ana committed:
updated, may kill testnet but yolo

dr7ana committed:
yay clients connecting

dr7ana committed:
yay path building

dr7ana committed:
squash, but more the zucchini type

dr7ana committed:
maybe a little green beans too

dr7ana committed:
any beets on the side?

dr7ana committed:
whatsup paths

dr7ana committed:
Better linux fd poller
